164402 - Andrey Vinicius Santos Souza

In [ ]:
import numpy as np
import sympy as sp

np.set_printoptions(precision=8, suppress=True)

def lu_com_pivotamento_parcial(A):
    n = A.shape[0]
    L = np.eye(n)
    U = A.copy().astype(np.float64)
    P = np.eye(n)

    for i in range(n):
        # Pivotamento parcial
        max_row = np.argmax(np.abs(U[i:, i])) + i
        if i != max_row:
            U[[i, max_row], :] = U[[max_row, i], :]
            P[[i, max_row], :] = P[[max_row, i], :]
            if i > 0:
                L[[i, max_row], :i] = L[[max_row, i], :i]

        for j in range(i+1, n):
            if U[i, i] == 0:
                raise ZeroDivisionError("Pivô zero encontrado!")
            fator = U[j, i] / U[i, i]
            L[j, i] = fator
            U[j, :] -= fator * U[i, :]

    return P, L, U

def resolver_sistema_lu(P, L, U, b):
    # Aplica permutação ao vetor b
    Pb = P @ b

    # Substituição progressiva para resolver L*y = Pb
    n = len(b)
    y = np.zeros(n)
    for i in range(n):
        y[i] = Pb[i] - np.dot(L[i, :i], y[:i])

    # Substituição regressiva para resolver U*x = y
    x = np.zeros(n)
    for i in range(n-1, -1, -1):
        x[i] = (y[i] - np.dot(U[i, i+1:], x[i+1:])) / U[i, i]

    return x

def entrada_matriz():
    n = int(input("Digite o tamanho da matriz no formato 'n' para nxn: "))
    A = np.zeros((n, n))
    b = np.zeros(n)

    print("\n📥 Preenchendo a matriz A:")
    for i in range(n):
        for j in range(n):
            A[i, j] = float(input(f"  a{i+1}{j+1}: "))

    print("\n📥 Preenchendo o vetor b:")
    for i in range(n):
        b[i] = float(input(f"  b{i+1}: "))

    return A, b

# Execução principal
A, b = entrada_matriz()
P, L, U = lu_com_pivotamento_parcial(A)
x = resolver_sistema_lu(P, L, U, b)

# Exibir resultados com 8 casas decimais
print("\n✅ Matriz A:")
display(sp.Matrix(A))

print("\n🔄 Matriz de Permutação P:")
display(sp.Matrix(P))

print("\n🟩 Matriz L (8 casas decimais):")
display(sp.Matrix(np.round(L, 8)))

print("\n🟥 Matriz U (8 casas decimais):")
display(sp.Matrix(np.round(U, 8)))

print("\n📌 Vetor solução x (8 casas decimais):")
for i, val in enumerate(x):
    print(f"x{i+1} = {val:.8f}")

# Verificação final
print("\n🔍 Verificação A * x:")
Ax = A @ x
display(sp.Matrix(np.round(Ax, 8)))
print("\n🔍 Vetor b original:")
display(sp.Matrix(np.round(b, 8)))


Digite o tamanho da matriz no formato 'n' para nxn: 3

📥 Preenchendo a matriz A:
  a11: 3
  a12: 1
  a13: -1
  a21: 2
  a22: 4
  a23: 1
  a31: 1
  a32: -2
  a33: 5

📥 Preenchendo o vetor b:
  b1: 500
  b2: 800
  b3: 300

✅ Matriz A:


Matrix([
[3.0,  1.0, -1.0],
[2.0,  4.0,  1.0],
[1.0, -2.0,  5.0]])


🔄 Matriz de Permutação P:


Matrix([
[1.0, 0.0, 0.0],
[0.0, 1.0, 0.0],
[0.0, 0.0, 1.0]])


🟩 Matriz L (8 casas decimais):


Matrix([
[       1.0,  0.0, 0.0],
[0.66666667,  1.0, 0.0],
[0.33333333, -0.7, 1.0]])


🟥 Matriz U (8 casas decimais):


Matrix([
[3.0,        1.0,       -1.0],
[0.0, 3.33333333, 1.66666667],
[0.0,        0.0,        6.5]])


📌 Vetor solução x (8 casas decimais):
x1 = 155.38461538
x2 = 104.61538462
x3 = 70.76923077

🔍 Verificação A * x:


Matrix([
[500.0],
[800.0],
[300.0]])


🔍 Vetor b original:


Matrix([
[500.0],
[800.0],
[300.0]])

# **Questão** **3**.**b**

Os mesmos valores *x*, *y* e *z* utilizando Lower and Upper  **são**:

📌 Vetor solução (8 casas decimais):

**x** = 155.38461538

**y** = 104.61538462

**z** = 70.76923077

In [ ]:
import numpy as np
import sympy as sp

np.set_printoptions(precision=12, suppress=True)

def lu_com_pivotamento_parcial(A):
    n = A.shape[0]
    L = np.eye(n)
    U = A.copy().astype(np.float64)
    P = np.eye(n)

    for i in range(n):
        # Pivotamento parcial
        max_row = np.argmax(np.abs(U[i:, i])) + i
        if i != max_row:
            U[[i, max_row], :] = U[[max_row, i], :]
            P[[i, max_row], :] = P[[max_row, i], :]
            if i > 0:
                L[[i, max_row], :i] = L[[max_row, i], :i]

        for j in range(i+1, n):
            if U[i, i] == 0:
                raise ZeroDivisionError("Pivô zero encontrado!")
            fator = U[j, i] / U[i, i]
            L[j, i] = fator
            U[j, :] -= fator * U[i, :]

    return P, L, U

def resolver_sistema_lu(P, L, U, b):
    # Aplica permutação ao vetor b
    Pb = P @ b

    # Substituição progressiva para resolver L*y = Pb
    n = len(b)
    y = np.zeros(n)
    for i in range(n):
        y[i] = Pb[i] - np.dot(L[i, :i], y[:i])

    # Substituição regressiva para resolver U*x = y
    x = np.zeros(n)
    for i in range(n-1, -1, -1):
        x[i] = (y[i] - np.dot(U[i, i+1:], x[i+1:])) / U[i, i]

    return x

def entrada_matriz():
    n = int(input("Digite o tamanho da matriz no formato 'n' para nxn: "))
    A = np.zeros((n, n))
    b = np.zeros(n)

    print("\n📥 Preenchendo a matriz A:")
    for i in range(n):
        for j in range(n):
            A[i, j] = float(input(f"  a{i+1}{j+1}: "))

    print("\n📥 Preenchendo o vetor b:")
    for i in range(n):
        b[i] = float(input(f"  b{i+1}: "))

    return A, b

# Execução principal
A, b = entrada_matriz()
P, L, U = lu_com_pivotamento_parcial(A)
x = resolver_sistema_lu(P, L, U, b)

# Exibir resultados com 12 casas decimais
print("\n✅ Matriz A:")
display(sp.Matrix(A))

print("\n🔄 Matriz de Permutação P:")
display(sp.Matrix(P))

print("\n🟩 Matriz L (12 casas decimais):")
display(sp.Matrix(np.round(L, 12)))

print("\n🟥 Matriz U (12 casas decimais):")
display(sp.Matrix(np.round(U, 12)))

print("\n📌 Vetor solução x (12 casas decimais):")
for i, val in enumerate(x):
    print(f"x{i+1} = {val:.12f}")

# Verificação final
print("\n🔍 Verificação A * x:")
Ax = A @ x
display(sp.Matrix(np.round(Ax, 12)))
print("\n🔍 Vetor b original:")
display(sp.Matrix(np.round(b, 12)))


Digite o tamanho da matriz no formato 'n' para nxn: 3

📥 Preenchendo a matriz A:
  a11: 3
  a12: 1
  a13: -1
  a21: 2
  a22: 4
  a23: 1
  a31: 1
  a32: -2
  a33: 5

📥 Preenchendo o vetor b:
  b1: 500
  b2: 800
  b3: 300

✅ Matriz A:


Matrix([
[3.0,  1.0, -1.0],
[2.0,  4.0,  1.0],
[1.0, -2.0,  5.0]])


🔄 Matriz de Permutação P:


Matrix([
[1.0, 0.0, 0.0],
[0.0, 1.0, 0.0],
[0.0, 0.0, 1.0]])


🟩 Matriz L (12 casas decimais):


Matrix([
[           1.0,  0.0, 0.0],
[0.666666666667,  1.0, 0.0],
[0.333333333333, -0.7, 1.0]])


🟥 Matriz U (12 casas decimais):


Matrix([
[3.0,            1.0,           -1.0],
[0.0, 3.333333333333, 1.666666666667],
[0.0,            0.0,            6.5]])


📌 Vetor solução x (12 casas decimais):
x1 = 155.384615384615
x2 = 104.615384615385
x3 = 70.769230769231

🔍 Verificação A * x:


Matrix([
[500.0],
[800.0],
[300.0]])


🔍 Vetor b original:


Matrix([
[500.0],
[800.0],
[300.0]])

# **Questão** **3**.**c**

Os mesmos valores *x*, *y* e *z* utilizando Lower and Upper  **são**:

📌 Vetor solução (12 casas decimais):

**x** = 155.384615384615

**y** = 104.615384615385

**z** = 70.769230769231